**LESSON 15: HANDS-ON PRACTICE**
=

*EXERCISE ONE*
=

*Sales Performance Calculator (Business Analytics)*
=

*1. Create Dataset*
=

In [9]:
# Import NumPy under the conventional alias "np"
import numpy as np

In [14]:
# Daily sales figures (in Naira) for one week, stored as a 1-D NumPy array
daily_sales = np.array([125000, 150000, 175000,140000, 190000, 210000, 160000])

# Inspect the array: its values, data type (dtype) and shape (7 elements in 1 dimension)
print(f"""
Daily Sales: {daily_sales}
Type: {daily_sales.dtype}
shape: {daily_sales.shape}
""")


Daily Sales: [125000 150000 175000 140000 190000 210000 160000]
Type: int64
shape: (7,)



*2. Calculate Total Sales*
=

In [15]:
# np.sum() adds up all elements in the array to give the week's total sales
total_sales = np.sum(daily_sales)

print(f"Total Sales: ₦{total_sales}")

Total Sales: ₦1150000


*3. Calculate Average Daily Sales*
=

In [16]:
# np.mean() returns the average daily sales across the week
average_sales = np.mean(daily_sales)

# :.2f formats the result to 2 decimal places
print(f"Average Sales: ₦{(average_sales):.2f}")

Average Sales: ₦164285.71


*4. Sales Adjustment*
=

In [20]:
# Multiplying an array by a scalar applies it to every element (broadcasting),
# so this computes a 10% increment for each day
sales_increment = (daily_sales) * 0.1
# Element-wise addition gives the adjusted (increased by 10%) sales per day
adjusted_sales = daily_sales + sales_increment

print(f"Adjusted Sales: {adjusted_sales}")

Adjusted Sales: [137500. 165000. 192500. 154000. 209000. 231000. 176000.]


*5. Compare Original and Adjusted Sales*
=

In [21]:
# Element-wise subtraction shows how much each day's sales changed after the adjustment
# (this will equal the 10% increment calculated above)
sales_difference = adjusted_sales - daily_sales

print(f"Difference in Sales: {sales_difference}")

Difference in Sales: [12500. 15000. 17500. 14000. 19000. 21000. 16000.]


*6. Mathematical Functions*
=

In [24]:
# Re-compute the two summary statistics together using NumPy's math functions
average_sales = np.mean(daily_sales)   # average daily sales
total_sales = np.sum(daily_sales)      # total sales for the week

print(f"Average Sales: ₦{(average_sales):.2f}")
print(f"Total Sales: ₦{total_sales}")

Average Sales: ₦164285.71
Total Sales: ₦1150000


The average daily sales figure shows that the average of the totals sales for a week is centralized around ₦164285.71, that is in each a day, there should be a sales of that amount/close to that amount for the business to have a total sales of ₦1150000 at the end of the week.

*EXERCISE TWO*
=

*Student Performance Analysis (Eduction Analytics)*
=

*1. Create and Inspect Data*
=

In [25]:
# Re-import NumPy for this exercise (already imported earlier, so this is optional)
import numpy as np

In [26]:
# Test scores for 10 students, stored as a NumPy array
scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])

# Inspect the array: its values, data type and shape
print(f"""
Scores: {scores}
Type: {scores.dtype}
shape: {scores.shape}
""")


Scores: [62 75 81 69 88 94 73 85 77 91]
Type: int64
shape: (10,)



*2. Calculate Mean Score*
=

In [27]:
# Mean (average) score of the class
mean_score = np.mean(scores)

print(f"Mean Score: {mean_score}")

Mean Score: 79.5


*3. Calculate Each Student's Deviaion*
=

In [30]:
# Deviation = how far each student's score is from the mean
# (positive = above average, negative = below average)
deviations = scores - mean_score

print(f"Student Deviations: {deviations}")

Student Deviations: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]


*4. Identify Positive and Negative Deviations*
=

- The positive deviation means the student score is greater than the "mean" score of the total students scores
- The negative deviation means the student score is less than the "mean" score of the total students scores
- When the deviation value is close to zero, its means the student score is almost close to the "mean" score of the total students scores

*5. Calculate Squared Deviations*
=

In [31]:
# Square each deviation element-wise so negatives don't cancel out positives
squared_deviation = (deviations) **2

print(f"Squared Deviation: {squared_deviation}")

Squared Deviation: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]


*6. Calculate Sum of Squared Deviations*
=

In [32]:
# Add up all the squared deviations (numerator of the variance formula)
sum_squared_deviations = np.sum(squared_deviation)

print(f"Sum of Squared Deviations: {sum_squared_deviations}")

Sum of Squared Deviations: 932.5


*7. Calculate Standard Deviation*
=

In [36]:
# Sample variance divides by (n - 1) rather than n (Bessel's correction),
# which is the standard choice when the data is a sample of a larger population
denominator = len(scores) - 1
variance = (sum_squared_deviations) / denominator
# Standard deviation is the square root of the variance
# (same units as the original scores)
standard_deviation = np.sqrt(variance)

print(f"""
Variance: {(variance):.2f}
Standard Deviation: {(standard_deviation):.2f}
""")


Variance: 103.61
Standard Deviation: 10.18



*EXERCISE THREE*
=

*Trigonometric Series and Convergence (Engineering & Scientific Computing)*
=

In [37]:
# Re-import NumPy for this exercise (already imported earlier, so this is optional)
import numpy as np

*1. Define Angle*
=

In [40]:
# NumPy's trig functions expect angles in radians, not degrees,
# so convert 30 degrees to radians first
theta = 30

theta_radians = np.radians(theta)

print(f"Angle in Radian: {theta_radians}")

Angle in Radian: 0.5235987755982988


*2. Create Sequence of Terms*
=

In [50]:
# np.arange(start, stop, step) creates the term numbers 1, 2, ..., 99
# (the stop value, 100, is excluded)
k = np.arange(1, 100, 1)

print(k)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]


*3. Calculate Trigonometric Component*
=

In [59]:
# Compute sin(k * theta) for every term number k in a single vectorised operation
sine_function_component = np.sin(k * theta_radians)

*4. Build Series Terms*
=

In [62]:
# Divide each sine value by its term number k to build the series terms: sin(k*theta) / k
sine_component_division = np.sin(k * theta_radians) / k

*5. Calculate Sum*
=

In [63]:
# Sum all 99 terms of the series
sum_sine_component = np.sum(sine_component_division)

print(sum_sine_component)

1.3136674864204645


*6. Increase Number of Terms*
=

In [68]:
# Repeat the same calculation with more terms (1 to 999) to see how the sum changes
a = np.arange(1, 1000, 1)
sine_function_component = np.sin(a * theta_radians)
sine_component_division = np.sin(a * theta_radians) / a
sum_sine_component = np.sum(sine_component_division)

print(sum_sine_component)

1.3094937000502784


In [69]:
# Repeat again with even more terms (1 to 9999); the sum settles closer to a fixed value
b = np.arange(1, 10000, 1)
sine_function_component = np.sin(b * theta_radians)
sine_component_division = np.sin(b * theta_radians) / b
sum_sine_component = np.sum(sine_component_division)

print(sum_sine_component)

1.3090469066682815


*7. Investigate Convergence*
=

        Number of Terms        |   Series Result
    ___________________________|______________________
    1.        99               |  1.3136674864204645
    2.        999              |  1.3094937000502784
    3.        9999             |  1.3090469066682815

    As the number of terms increases, the  sum of the sine functions converges to 1, thats as the number of terms increases the sum reduces 

*8. Explain Convergence*
=

According to the mathematical experiment performed in this script, Convergence can simply be described as the procedure carried out in which series of terms converges to a value as there exist an increase in the number of terms.

1. When fewer terms is used, the result of the sum of series tends to be greater value
2. When the number of terms is increased, as the inrement continues, the resulting sum get smaller and systematically converges to a value
3. Increasing the number of terms changes the approximation because, the increment contains larger number which are used to divide the constant numerator value
4. When the result is getting closer to a limit value, it simply means the resulting sum is converging. 

**BONUS CHALLENGE**
=

*1. Project Title*
=

Stock Market Volatility & Moving Average Analysis

*2. Scenario*
=

Finance: A financial analyst wants to analyze the daily closing prices of a fictional stock over a 7-day trading week to determine its average price, total volume traded, and daily fluctuations.

*3. Objective*
=

To use NumPy to calculate key financial metrics (sum, mean, and volatility) from raw stock data, demonstrating how arrays can simplify mathematical operations on datasets.

*4. Dataset*
=

A NumPy array representing the daily closing price of a stock from Monday to Friday (plus a weekend adjustment), and a sequence representing the trading days.

In [7]:
import numpy as np

# 1. Create the Numerical Dataset
# Daily closing prices of a fictional stock for 7 days
# [Requirement met: np.array()]
stock_prices = np.array([150.25, 152.10, 149.50, 153.75, 155.00, 151.25, 154.50])

# Create a sequence representing the days (Day 1 to Day 7)
# [Requirement met: np.arange()]
days = np.arange(1, 8)

print(f"Stock Data: {stock_prices}")
print(f"Days: {days}")
print(f"Shape: {stock_prices.shape}")
print("-" * 30)

# 2. Perform Calculations

# Calculation A: Total sum of prices (Total value of portfolio if bought 1 share each day)
# [Requirement met: np.sum()]
total_value = np.sum(stock_prices)

# Calculation B: Average price over the week
# [Requirement met: np.mean()]
average_price = np.mean(stock_prices)

# Calculation C: Calculate the daily price changes (Difference from previous day)
# Using np.subtract() to find the difference between consecutive days
# We slice the array: stock_prices[1:] (days 2-7) and stock_prices[:-1] (days 1-6)
# [Requirement met: np.subtract()]
daily_changes = np.subtract(stock_prices[1:], stock_prices[:-1])

# Calculation D: Find the magnitude of volatility (absolute changes)
# [Requirement met: Element-wise array operations]
absolute_changes = np.abs(daily_changes)

# Calculation E: Calculate the Square Root of the variance (Standard Deviation) manually
# First, find the deviation of each day from the mean
deviation = np.subtract(stock_prices, average_price)
# Square the deviations (element-wise)
squared_dev = np.multiply(deviation, deviation)
# Find the variance (mean of squared deviations)
# [Requirement met: np.mean()]
variance = np.mean(squared_dev)
# Standard Deviation (Volatility)
# [Requirement met: np.sqrt()]
volatility = np.sqrt(variance)

# 3. Output the Results
print(f"Total Value of Stock: ${total_value:.2f}")
print(f"Average Stock Price: ${average_price:.2f}")
print(f"Daily Price Changes: {daily_changes}")
print(f"Volatility (Std Dev): ${volatility:.2f}")

# 4. Interpretation
print("\n--- INTERPRETATION ---")
print(f"The stock closed at an average of ${average_price:.2f} over the 7 days.")
print(f"The total sum of the closing prices was ${total_value:.2f}.")
print(f"The daily changes show the stock fluctuated by an average of ${np.mean(absolute_changes):.2f} per day.")
print(f"A volatility (standard deviation) of ${volatility:.2f} indicates moderate risk; "
      f"the price typically stays within ${average_price - volatility:.2f} to ${average_price + volatility:.2f}.")

Stock Data: [150.25 152.1  149.5  153.75 155.   151.25 154.5 ]
Days: [1 2 3 4 5 6 7]
Shape: (7,)
------------------------------
Total Value of Stock: $1066.35
Average Stock Price: $152.34
Daily Price Changes: [ 1.85 -2.6   4.25  1.25 -3.75  3.25]
Volatility (Std Dev): $1.98

--- INTERPRETATION ---
The stock closed at an average of $152.34 over the 7 days.
The total sum of the closing prices was $1066.35.
The daily changes show the stock fluctuated by an average of $2.82 per day.
A volatility (standard deviation) of $1.98 indicates moderate risk; the price typically stays within $150.36 to $154.31.
